File: go/tencent/array_and_str/three_sum.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
}

func threeSum(nums []int) [][]int {
	res := make([][]int, 0)
	if nums == nil || len(nums) < 3 {
		return res
	}
	sort.Ints(nums)
	// 跳过相同的数字 防止重复
	for i, num := range nums {
		if canGenerate(nums, i) {
			for _, twoNumberArray := range findTwoSum(nums, -num, i+1) {
				if len(twoNumberArray) > 0 {
					res = append(res, []int{num, twoNumberArray[0], twoNumberArray[1]})
				}
			}
		}
	}
	return res
}

func canGenerate(nums []int, start int) bool {
	if start > 0 && nums[start-1] == nums[start] {
		return false
	}
	return true
}

// 用之前的 map 的方式做 可能会存在重复 因此这儿用二分来做
func findTwoSum(nums []int, target int, start int) [][]int {
	res := make([][]int, 0)
	i, j := start, len(nums)-1
	for i < j {
		tempSum := nums[i] + nums[j]
		if tempSum > target {
			j--
		} else if tempSum < target {
			i++
		} else {
			// 相等 需要跳过相等的
			res = append(res, []int{nums[i], nums[j]})
			for i < j && nums[i] == nums[i+1] {
				i++
			}
			// 跳到下一个
			i++

			for j > i && nums[j] == nums[j-1] {
				j--
			}
			j--
		}
	}
	return res
}

File: java/leetcode-explore/highFrequency/array/ThreeSum.java
package highFrequency.array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {

    // 既然不能用 recursion 也就意味着
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length; i++) {
            // 跳过相等的数字
            if (i != 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int target = -nums[i];

            for (int l = i + 1, r = nums.length - 1; l < r; ) {
                int sum = nums[l] + nums[r];

                if (sum > target) {
                    r--;
                } else if (sum < target) {
                    l++;
                } else {
                    // 等于
                    int finalI = i;
                    int finalL = l;
                    int finalR = r;
                    res.add(new ArrayList<>() {{
                        add(nums[finalI]);
                        add(nums[finalL]);
                        add(nums[finalR]);
                    }});
                    // 跳过相等的数字
                    do {
                        l++;
                    } while (l < r && nums[l] == nums[l - 1]);
                    do {
                        r--;
                    } while (l < r && nums[r] == nums[r + 1]);
                }
            }
        }

        return res;
    }

    // recursion 超时
    public List<List<Integer>> threeSumWithRecursion(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();

        Arrays.sort(nums);
        recursion(res, new ArrayList<>(), 0, nums, 0);
        return res;
    }

    public void recursion(List<List<Integer>> res, List<Integer> tmp, int start, int[] nums, int target) {
        if (tmp.size() == 3) {
            if (target == 0) {
                res.add(new ArrayList<>(tmp));
            }
            return;
        }

        for (int i = start; i < nums.length; i++) {
            if (canGenerate(nums, start, i)) {
                tmp.add(nums[i]);
                recursion(res, tmp, i + 1, nums, target - nums[i]);
                tmp.remove(tmp.size() - 1);
            }
        }
    }

    public boolean canGenerate(int[] nums, int start, int end) {
        for (int i = start; i < end; i++) {
            if (nums[i] == nums[end]) {
                return false;
            }
        }
        return true;
    }
}
